lib/quicrq.c

```diff
@@ -1788,6 +1789,101 @@ int quicrq_receive_stream_data(quicrq_stream_ctx_t* stream_ctx, uint8_t* bytes,
+                    quicrq_message_t incoming = { 0 };
+                    const uint8_t* r_bytes = quicrq_msg_decode(uni_stream_ctx->message_buffer.buffer, uni_stream_ctx->message_buffer.buffer + stream_ctx->message_receive.message_size, &incoming);
+
+                    if (r_bytes == NULL) {
+                        /* Message was incorrect */
+                        ret = -1;
+                    } else {
+                        switch (incoming.message_type) {
+                            case QUICRQ_ACTION_WARP_HEADER:
+                                uni_stream_ctx->current_group_id = incoming.group_id;
+                                uni_stream_ctx->receive_state = quicrq_receive_warp_header;
+                                quicrq_stream_ctx_t* ctrl_stream_ctx = quicrq_get_control_stream_for_media_id(cnx_ctx, incoming.media_id);
+                                uni_stream_ctx->receive_state = quicrq_receive_warp_header;
+                                uni_stream_ctx->control_stream_ctx = ctrl_stream_ctx;
+                                /* todo: add check for null control stream */
+                                quicrq_log_message(cnx_ctx, "UniStream %" PRIu64 ", received warp header message uni_id= %" PRIu64,
+                                                   ", ControlStream id= %" PRIu64 " media id=%" PRIu64,
+                                                   uni_stream_ctx->stream_id, uni_stream_ctx->control_stream_ctx->stream_id, incoming.media_id);
+
+                                break;
+                            case QUICRQ_ACTION_OBJECT_HEADER:
+                                if (uni_stream_ctx->receive_state != quicrq_receive_warp_header) {
+                                    /* Protocol error */
+                                    ret = -1;
+                                }
+                                else {
+                                    uni_stream_ctx->receive_state = quicrq_receive_object_header;
+                                    quicrq_stream_ctx_t* ctrl_stream_ctx = quicrq_get_control_stream_for_media_id(cnx_ctx, incoming.media_id);
+                                    uni_stream_ctx->control_stream_ctx = ctrl_stream_ctx;
+                                    /* Pass the repair data to the media consumer. */
+                                    ret = ctrl_stream_ctx->consumer_fn(quicrq_media_datagram_ready, ctrl_stream_ctx->media_ctx, picoquic_get_quic_time(ctrl_stream_ctx->cnx_ctx->qr_ctx->quic),
+                                                                  incoming.data, uni_stream_ctx->current_group_id, incoming.object_id,
+                                                                  incoming.offset, 0, incoming.flags, incoming.nb_objects_previous_group,
+                                                                  incoming.is_last_fragment, incoming.length);
+                                    ret = quicrq_cnx_handle_consumer_finished(ctrl_stream_ctx, 0, 0, ret);
```

Comment:
AFAIK, there is no "incoming offset", unless you plan doing a zero copy implementation. Offset will always be zero. Similarly, "is_last_fragment" will always be true.

Reply:
Did you encode the flags in the object header? Or the number of objects in previous group?

Reply:
yes I did encode it in the proto.c